Segfault error under Linux

Hi - I'm trying box within a docker container and getting errors when trying to import any R file (it works fine for importing libraries). It's any R file - even a 4-liner with one single function.

Please see below the error I'm getting. The base container is rocker/verse:3.6.1.

(I also installed it directly into the container through devtools::install_github - same error. For what its worth I put the install log at the bottom.)

Have you come across such a limitation/error?

Here's the container setup:
```
docker run --rm -it  rocker/verse:3.6.1 /bin/bash
```

Then within the container I start R and execute the following commands:
```
install.packages('devtools')
devtools::install_github('klmr/box@dev')
box::use(test=./test[...])
```
test is a simple file with one command to be exported (can be anything really - even if it's empty it's the same error).

The errors (first one is from the test example from above - the second one is the original error from my code - both are segfault errors):

```
Traceback:
 1: `$.box$ns`(ns, ".__module__.")
 2: ns$.__module__.
 3: assign(which, value, envir = ns$.__module__.)
 4: `namespace_info<-`(`*tmp*`, "exports", value = parse_export_specs(info,     exprs, mod_ns))
 5: load_from_source(info, mod_ns)
 6: `load_mod.box$mod_info`(info)
 7: load_mod(info)
 8: load_and_register(spec, info, caller)
 9: (function (declaration, alias, caller) {    if (identical(declaration, quote(expr = )) && identical(alias,         ""))         return()    spec = pa
rse_spec(declaration, alias)    info = rethrow_on_error(find_mod(spec, caller), sys.call(-1L))    load_and_register(spec, info, caller)})(dots[[1L]][[1L]
], dots[[2L]][[1L]], dots[[3L]][[1L]])
10: mapply(FUN = f, ..., SIMPLIFY = FALSE)
11: Map(.f, ..., USE.NAMES = FALSE)
12: map(use_one, imports, aliases, list(caller))
13: box::use(test = ./rsource/test[...])
```

```
 *** caught segfault ***
address 0x4844e190, cause 'memory not mapped'

Traceback:
 1: `$.box$ns`(ns, ".__module__.")
 2: ns$.__module__.
 3: get0(which, ns$.__module__., inherits = FALSE, ifnotfound = default)
 4: namespace_info(calling_ns, "info")
 5: dirname(namespace_info(calling_ns, "info")$source_path)
 6: calling_mod_path(caller)
 7: map(file.path, base_paths, list(c(simple_mod, nested_mod)))
 8: mod_file_candidates(spec, base_paths)
 9: find_in_path(spec, calling_mod_path(caller))
10: find_local_mod(spec, caller)
11: `find_mod.box$mod_spec`(spec, caller)
12: find_mod(spec, caller)
13: doTryCatch(return(expr), name, parentenv, handler)
14: tryCatchOne(expr, names, parentenv, handlers[[1L]])
15: tryCatchList(expr, classes, parentenv, handlers)
16: tryCatch(expr, error = function(error) rethrow(error, call))
17: rethrow_on_error(find_mod(spec, caller), sys.call(-1L))
18: (function (declaration, alias, caller) {    if (identical(declaration, quote(expr = )) && identical(alias,         ""))         return()    spec = pa
rse_spec(declaration, alias)    info = rethrow_on_error(find_mod(spec, caller), sys.call(-1L))    load_and_register(spec, info, caller)})(dots[[1L]][[1L]
], dots[[2L]][[1L]], dots[[3L]][[1L]])
19: mapply(FUN = f, ..., SIMPLIFY = FALSE)
20: Map(.f, ..., USE.NAMES = FALSE)
21: map(use_one, imports, aliases, list(caller))
22: box::use(./system[...])
23: eval(exprs, mod_ns)
24: eval(exprs, mod_ns)
25: load_from_source(info, mod_ns)
26: `load_mod.box$mod_info`(info)
27: load_mod(info)
28: load_and_register(spec, info, caller)
29: (function (declaration, alias, caller) {    if (identical(declaration, quote(expr = )) && identical(alias,         ""))         return()    spec = pa
rse_spec(declaration, alias)    info = rethrow_on_error(find_mod(spec, caller), sys.call(-1L))    load_and_register(spec, info, caller)})(dots[[1L]][[1L]
], dots[[2L]][[1L]], dots[[3L]][[1L]])
30: mapply(FUN = f, ..., SIMPLIFY = FALSE)
31: Map(.f, ..., USE.NAMES = FALSE)
32: map(use_one, imports, aliases, list(caller))
33: box::use(sharedHelpers/PackageEntry[setParam, createNewParameterObj])
```
#
```
> devtools::install_github("klmr/box@dev")
Downloading GitHub repo klmr/box@dev
✔  checking for file ‘/tmp/Rtmpn2I1Xz/remotes4f276765c7/klmr-box-b0750e4/DESCRIPTION’ ...
─  preparing ‘box’:
✔  checking DESCRIPTION meta-information ...
─  cleaning src
─  checking for LF line-endings in source and make files and shell scripts
─  checking for empty or unneeded directories
─  building ‘box_1.0.2.9000.tar.gz’

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
* installing *source* package ‘box’ ...
** using staged installation
** libs
gcc -I"/usr/local/lib/R/include" -DNDEBUG   -I/usr/local/include  -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_
FORTIFY_SOURCE=2 -g  -c exports.c -o exports.o
gcc -I"/usr/local/lib/R/include" -DNDEBUG   -I/usr/local/include  -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_
FORTIFY_SOURCE=2 -g  -c lookup.c -o lookup.o
In file included from lookup.c:2:0:
lookup.c: In function ‘strict_extract’:
lookup.c:24:25: warning: implicit declaration of function ‘Rf_installTrChar’ [-Wimplicit-function-declaration]
     SEXP name = PROTECT(Rf_installTrChar(STRING_ELT(e2, 0)));
                         ^
/usr/local/lib/R/include/Rinternals.h:852:31: note: in definition of macro ‘PROTECT’
 #define PROTECT(s) Rf_protect(s)
                               ^
lookup.c:24:25: warning: passing argument 1 of ‘Rf_protect’ makes pointer from integer without a cast [-Wint-conversion]
     SEXP name = PROTECT(Rf_installTrChar(STRING_ELT(e2, 0)));
                         ^
/usr/local/lib/R/include/Rinternals.h:852:31: note: in definition of macro ‘PROTECT’
 #define PROTECT(s) Rf_protect(s)
                               ^
In file included from lookup.c:2:0:
/usr/local/lib/R/include/Rinternals.h:1616:6: note: expected ‘SEXP {aka struct SEXPREC *}’ but argument is of type ‘int’
 SEXP Rf_protect(SEXP);
      ^~~~~~~~~~
gcc -I"/usr/local/lib/R/include" -DNDEBUG   -I/usr/local/include  -fpic  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_
FORTIFY_SOURCE=2 -g  -c unlock_env.c -o unlock_env.o
gcc -shared -L/usr/local/lib/R/lib -L/usr/local/lib -o box.so exports.o lookup.o unlock_env.o -L/usr/local/lib/R/lib -lR
installing to /usr/local/lib/R/site-library/00LOCK-box/00new/box/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (box)
```

--- a/NEWS.md
+++ b/NEWS.md
@@ -1,5 +1,7 @@
 # box (development version)
 
+* Fix: prevent segfault in R ≤ 3.6.1 caused by missing declaration of internal R
+    symbol (#213)
 * Fix: Allow exporting modules that were previously imported using a different
     prefix (#211)
 * Enhancement: Add standard module for core R packages (#200)

--- a/src/lookup.c
+++ b/src/lookup.c
@@ -1,6 +1,12 @@
 #define R_NO_REMAP
 #include "Rinternals.h"
 #include "R_ext/Parse.h"
+#include "Rversion.h"
+
+#if R_VERSION < R_Version(3, 6, 2)
+// Declaration is internal before R 3.6.2.
+SEXP Rf_installTrChar(SEXP);
+#endif
 
 static SEXP parent_frame(void);
 static SEXP sys_call(SEXP parent);

